From ac22ce1947f4a845b7e860bf762c902996fe832d Mon Sep 17 00:00:00 2001
From: Daniel Micay <daniel.micay@grapheneos.org>
Date: Fri, 2 Feb 2024 18:54:40 -0500
Subject: [PATCH 01/24] run full explicit GC in SystemUI and system_server
 after locking

ART has special support for System.gc() and Android already uses this to
purge memory after unlocking. It makes sense to purge as much as we can
when locking too.
---
 .../com/android/systemui/keyguard/KeyguardViewMediator.java | 3 +++
 .../server/policy/keyguard/KeyguardServiceDelegate.java     | 6 ++++++
 2 files changed, 9 insertions(+)

diff --git a/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java b/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
index cf7475590..942dc489e 100644
--- a/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
+++ b/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
@@ -3293,6 +3293,9 @@ public class KeyguardViewMediator implements CoreStartable,
             userActivity();
             mUpdateMonitor.setKeyguardGoingAway(false);
             mKeyguardViewControllerLazy.get().setKeyguardGoingAwayState(false);
+            System.gc();
+            System.runFinalization();
+            System.gc();
             mShowKeyguardWakeLock.release();
         }
         mKeyguardDisplayManager.show();
diff --git a/services/core/java/com/android/server/policy/keyguard/KeyguardServiceDelegate.java b/services/core/java/com/android/server/policy/keyguard/KeyguardServiceDelegate.java
index 2d25ee154..94d9d7dc4 100644
--- a/services/core/java/com/android/server/policy/keyguard/KeyguardServiceDelegate.java
+++ b/services/core/java/com/android/server/policy/keyguard/KeyguardServiceDelegate.java
@@ -217,6 +217,12 @@ public class KeyguardServiceDelegate {
     public void onShowingStateChanged(boolean showing) {
         mKeyguardReportedState.showing = showing;
         mCallback.onShowingChanged();
+
+        if (showing) {
+            System.gc();
+            System.runFinalization();
+            System.gc();
+        }
     }
 
     /**
-- 
2.55.0

